### PR TITLE
*: wrap etcd transaction with client context.

### PR DESCRIPTION
**Pull request by huachaohuang**
Address issue #190 .

Wrap `txn` and `context` in `slowLogTxn` and call `cancel()` on `Commit()`.

### CODE REVIEW

File: server/tso.go [hunk 1]
```diff
@@ -64,12 +63,7 @@ func (s *Server) saveTimestamp(now time.Time) error {
 	data := uint64ToBytes(uint64(now.UnixNano()))
 	key := s.getTimestampPath()
 
-	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
-	resp, err := s.slowLogTxn(ctx).
-		If(s.leaderCmp()).
-		Then(clientv3.OpPut(key, string(data))).
-		Commit()
-	cancel()
+	resp, err := s.leaderTxn().Then(clientv3.OpPut(key, string(data))).Commit()
```
Comment (qiuyesuifeng) [inline server/tso.go]: Should we call cancel function here?

Comment (huachaohuang) [inline server/tso.go]: No, `slowLogTxn` will call `cancel()` on `Commit()`.

File: server/util.go [hunk 2]
```diff
@@ -105,7 +114,7 @@ func (t *slowLogTxn) Commit() (*clientv3.TxnResponse, error) {
 	if cost := time.Now().Sub(start); cost > slowRequestTime {
```
Comment (siddontang) [inline server/util.go]: we can add a metric for time transaction.

File: server/util.go [hunk 3]
```diff
@@ -99,22 +99,50 @@ type slowLogTxn struct {
...
+		labels = append(labels, "slow")
 		log.Warnf("txn runs too slow, resp: %v, err: %v, cost: %s", resp, err, cost)
 	}
-	t.cancel()
+	if err != nil {
+		labels = append(labels, "failed")
+	} else {
+		labels = append(labels, "success")
+	}
+	for _, label := range labels {
```
Comment (qiuyesuifeng) [inline server/util.go]: Seems no need to use slice, directly set metrics is OK.

File: server/metrics.go [hunk 4]
```diff
@@ -71,6 +71,23 @@ var (
...
 			Help:      "Counter of balancers.",
 		}, []string{"result"})
+
+	txnCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "pd",
+			Subsystem: "txn",
+			Name:      "txns_count",
+			Help:      "Counter of txns.",
+		}, []string{"type"})
```
Comment (qiuyesuifeng) [inline server/metrics.go]: It is better to use result instead of type, for you only want to trace the result of txns.

**Comment by qiuyesuifeng** [comment]
LGTM

File: server/util.go [hunk 5]
```diff
@@ -96,15 +96,47 @@ func uint64ToBytes(v uint64) []byte {
...
 	start := time.Now()
 	resp, err := t.Txn.Commit()
-	if cost := time.Now().Sub(start); cost > slowRequestTime {
+	t.cancel()
+
+	cost := time.Now().Sub(start)
+	if cost > slowRequestTime {
 		log.Warnf("txn runs too slow, resp: %v, err: %v, cost: %s", resp, err, cost)
 	}
+	label := "success"
```
Comment (siddontang) [inline server/util.go]: define a const var

Comment (qiuyesuifeng) [inline server/util.go]: If we do this, it is better to modify the same `success` metrics in pd project.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**